content/4_Gold/Combinatorics.mdx

```diff
@@ -289,6 +289,108 @@ int main() {
 </CPPSection>
 </LanguageSection>
 
+### Derangements
+
+<FocusProblem problem="sample2" />
+
+The number of derangements of $n$ numbers, expressed as $!n$, is the number of permutations such that no element appears in its original position. Informally, it is the number of ways $n$ hats can be returned to $n$ people such that no person recieves their own hat. 
+
+### Method 1: Principle of Inclusion-Exclusion
+
+Suppose we had events $E_1, E_2, \dots, E_n$, where event $E_i$ corresponds to person $i$ recieving their own hat. We would like to calculate $n! - \lvert E_1 \cup E_2 \cup \dots \cup E_n \rvert$. 
+
+We subtract from $n!$ the number of ways for each event to occur; that is, consider the quantity $n! - \lvert E_1 \rvert - \lvert E_2 \rvert - \dots - \lvert E_n \rvert$. This undercounts, as we are subtracting cases where more than one event occurs too many times. Specifically, for a permutation where at least two events occur, we undercount by one. Thus, add back the number of ways for two events to occur. We can continue this process for every size of subsets of indices. The expression is now of the form:
+
+$$
+n! - \lvert E_1 \cup E_2 \cup \dots \cup E_n \rvert = \sum_{k = 1}^n (-1)^k \cdot (\text{number of permutations with $k$ fixed points})
+$$
+
+For a set size of $k$, the number of permutations with at least $k$ indicies can be computed by choosing a set of size $k$ that are fixed, and permuting the other indices. In mathematical terms:
+
+$$
+{n \choose k}(n-k)! = \frac{n!}{k!(n-k)!}(n-k)! = \frac{n!}{k!}
+$$
+
+Thus, the problem now becomes computing 
+
+$$
+n!\sum_{k=0}^n\frac{(-1)^k}{k!}
+$$
+
+which can be done in linear time.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+		
+// https://atcoder.github.io/ac-library/document_en/modint.html (included in grader)
+#include <atcoder/modint>
+using mint = atcoder::modint;
+
+int main() {
+	using namespace std;
```

Comment:
why on god's green earth is using namespace here

Reply:
if there's no function ambiguity in the main function, then it's fine to include namespace std

Reply:
nevermind that this is the only place in the entire codebase where it's here

Reply:
if this matters for you then change it